mlir/lib/Dialect/Rock/Transforms/ThreadwiseGemmLowering.cpp

```diff
@@ -248,88 +244,22 @@ struct XdlopsGemmV2RewritePattern : public OpConversionPattern<XdlopsGemmV2Op> {
-        //     argB = &matrixAElement[ki_i];
-        //     p_c_thread = mfma_type.template run<MPerXlops * mRepeats,
-        //                                         NPerXdlops * nRepeats,
-        //                                         AStride,
-        //                                         BStride>(argA, argB,
-        //       p_c_thread);
-        // }
-        int64_t outerLoopUpperBound = KPerThread;
-        int64_t KRepeats = KPack / k_base;
-
-        auto outerLoop = b.create<AffineForOp>(loc, 0, outerLoopUpperBound);
-        auto outerLoopb = ConversionPatternRewriter::atBlockBegin(
-            outerLoop.getBody(), b.getListener());
-        auto outerLoopiv = outerLoop.getInductionVar();
-
-        auto loadSingleKPack = [&](Value regOffset, Type elementType,
-                                   Value matrix) {
-          Value regIdx = outerLoopb.create<AddIOp>(loc, regOffset, outerLoopiv);
-          Value element = outerLoopb.create<memref::LoadOp>(
-              loc, elementType, matrix, ValueRange{regIdx});
-          return element;
-        };
-
-        Value matrixAElement =
-            loadSingleKPack(regAOffset, matrixAElementType, bufferA);
-        Value matrixBElement =
-            loadSingleKPack(regBOffset, matrixBElementType, bufferB);
-
-        auto innerLoop =
-            outerLoopb.create<AffineForOp>(loc, 0, KRepeats * k_base, k_base);
-        auto innerLoopb = ConversionPatternRewriter::atBlockBegin(
-            innerLoop.getBody(), outerLoopb.getListener());
-        auto innerLoopiv = innerLoop.getInductionVar();
-
-        // At this point, we are guaranteed that buffer element vectorization
-        // length (kPack) must be a multiple of k_base. Use extractsliceop
-        // to handle a independent data slice at a time.
-        Value argA = innerLoopb.create<ExtractSliceOp>(
-            loc, argType, matrixAElement, innerLoopiv);
-        Value argB = innerLoopb.create<ExtractSliceOp>(
-            loc, argType, matrixBElement, innerLoopiv);
-        populateMfma(innerLoopb, argA, argB, regCOffset);
+      auto mfmaLoop = b.create<TransformingForOp>(
+          loc, ArrayRef<ValueRange>{{zeroConstantOp}},
+          ArrayRef<Attribute>{b.getArrayAttr({})},
+          /*bounds=*/ArrayRef<int64_t>{kBasePerThread},
+          /*strides=*/ArrayRef<int64_t>{1},
+          /*forceUnroll=*/true, /*useIndexDiffs=*/true);
+      {
+        OpBuilder::InsertionGuard guard(b);
+        b.setInsertionPointToStart(mfmaLoop.getBody());
+        Value coord = mfmaLoop.getLowerCoords(/*domain=*/0)[0];
+
+        Value regIdxA = b.create<arith::AddIOp>(loc, coord, regAOffset);
+        Value regIdxB = b.create<arith::AddIOp>(loc, coord, regBOffset);
+        Value argA = b.create<memref::LoadOp>(loc, argType, bufferA, regIdxA);
+        Value argB = b.create<memref::LoadOp>(loc, argType, bufferB, regIdxB);
+        populateMfma(b, argA, argB, regCOffset);
```

Comment:
This used to be the part that I hate most to accommodate the incompatibility of kpack and kbase when they don't fit with each other. Now that you made the unit to be k_base, the loop is just natural. Good job!